src/crypto/NoteEncryption.cpp

```diff
@@ -0,0 +1,161 @@
+#include "NoteEncryption.hpp"
+#include "crypto/sha512.h"
+#include <stdexcept>
+#include "sodium.h"
+#include <boost/static_assert.hpp>
+#include "prf.h"
+#include "zerocash/Zerocash.h"
+
+#define NOTEENCRYPTION_CIPHER_KEYSIZE 32
+
+void clamp_curve25519(unsigned char *key)
+{
+    key[0] &= 248;
+    key[31] &= 127;
+    key[31] |= 64;
+}
+
+void KDF(unsigned char K[NOTEENCRYPTION_CIPHER_KEYSIZE],
+    const uint256 &dhsecret,
+    const uint256 &epk,
+    const uint256 &pk_enc,
+    const uint256 &seed,
```

Comment:
Let's call this hsig so it's clear what it is if you're looking at the code side-by-side with the protocol spec.